File: Hospital.c/src/paciente.c
#include "sistema.c"
#include "C:\Users\jhoan\Desktop\VScode\GitHub\HospitalManagementSystem\Hospital.c\include\paciente.h"

#define MaxNome 100
#define Leitos 25

union document
{
    char cpf[15];
    char rg[12];
};

struct pacientes
{
    char nome[MaxNome];
    Document documento;
    char enfermidade[50];
    char internado; // sim == 1, nao == 0
    char receita[100];
};

struct listapacientes
{
    Pacientes *pacientes;
    struct listapacientes *next;
    struct listapacientes *prev;
};

Listapacientes *lista_cria(void)
{
    Listapacientes *l = (Listapacientes *)malloc(sizeof(Listapacientes));
    return l;
}

Pacientes paciente_preenche(void)
{
    Pacientes paciente;
    char opcao_documento;

    printf("Digite o nome do paciente: ");
    scanf(" %[^\n]", paciente.nome);
    LimpaBuffer();
    corrige_nome(paciente.nome);

    printf("Digite a enfermidade do paciente: ");
    scanf(" %[^\n]s", paciente.enfermidade);
    LimpaBuffer();

    printf("Digite a receita para o paciente: ");
    scanf(" %[^\n]s", paciente.receita);
    LimpaBuffer();

    printf("Internado? (1-sim) (0-nao)\n");
    paciente.internado = LeOpcao('0', '1');

    printf("Digite '1' para cadastrar o CPF ou '2' para cadastrar o RG\n");
    opcao_documento = LeOpcao(OPCAO1, '2');
    if (opcao_documento == OPCAO1)
    {
        char cpf[15];
        lerCPF(cpf);
        strcpy(paciente.documento.cpf, cpf);
    }
    else if (opcao_documento == OPCAO2)
    {
        char rg[12];
        lerCPF(rg);
        strcpy(paciente.documento.rg, rg);
    }
    else
    {
        printf("Ocorreu um BUG\n");
        exit(1);
    }
    return paciente;
}

void paciente_add(Listapacientes *p, FILE *arquivo)
{
    Pacientes paciente = *(p->next->pacientes);
    fprintf(arquivo, "Nome: %s\n", paciente.nome);
    fprintf(arquivo, "Enfermidade: %s\n", paciente.enfermidade);
    fprintf(arquivo, "Receita: %s\n", paciente.receita);
    fprintf(arquivo, "Internado: %s\n", paciente.internado ? "Sim" : "Nao");
    if (strlen(paciente.documento.cpf) == 14)
    {
        fprintf(arquivo, "CPF: %s\n", paciente.documento.cpf);
    }
    else
    {
        fprintf(arquivo, "RG: %s\n", paciente.documento.rg);
    }
    fprintf(arquivo, "\n");
}

Listapacientes *lista_add(Listapacientes *l, Pacientes paciente)
{

    Listapacientes *p = l;
    while (p->next != NULL && strcmp(p->next->pacientes->nome, paciente.nome) < 0)
    {
        p = p->next;
    }

    Listapacientes *novo_no = (Listapacientes *)malloc(sizeof(Listapacientes));
    if (novo_no == NULL)
    {
        printf("Erro: memoria insuficiente.\n");
        exit(1);
    }
    novo_no->pacientes = (Pacientes *)malloc(sizeof(Pacientes));
    if (novo_no->pacientes == NULL)
    {
        printf("Erro: memoria insuficiente.\n");
        exit(1);
    }
    *novo_no->pacientes = paciente;

    novo_no->next = p->next;
    novo_no->prev = p;
    if (p->next != NULL)
    {
        p->next->prev = novo_no;
    }
    p->next = novo_no;
    return l;
}

Listapacientes *busca_paciente(Listapacientes *l, char nome[])
{
    Listapacientes *p;
    for (p = l; p != NULL; p = p->next)
    {
        if (strcmp(p->pacientes->nome, nome) == 0)
        {
            return p;
        }
    }
    return NULL;
}

void edita_paciente(Listapacientes *p)
{
    char opcao_documento;
    char cont;
    printf("\nNome: %s\n", p->pacientes->nome);
    printf("Enfermidade: %s\n", p->pacientes->enfermidade);
    printf("Receita: %s\n", p->pacientes->receita);
    printf("Internado: %s\n", p->pacientes->internado ? "Sim" : "Nao");
    if (strlen(p->pacientes->documento.cpf) == 14)
    {
        printf("CPF: %s\n", p->pacientes->documento.cpf);
    }
    else
    {
        printf("RG: %s\n", p->pacientes->documento.rg);
    }
    printf("Deseja mesmo editar o paciente acima? (1-sim)(2-nao)\n");
    cont = LeOpcao('1', '2');
    if (cont == '2')
    {
        printf("\n");
    }
    else if (cont == '1')
    {

        char opcao;
        printf("\nO que voce deseja editar?\n");
        printf("1 - Nome\n");
        printf("2 - Enfermidade\n");
        printf("3 - Receita\n");
        printf("4 - Internado\n");
        printf("5 - Documento\n");
        opcao = LeOpcao('1', '5');
        switch (opcao)
        {
        case OPCAO1:
            printf("\nDigite o novo nome: ");
            scanf(" %[^\n]s", p->pacientes->nome);
            LimpaBuffer();
            corrige_nome(p->pacientes->nome);
            printf("Paciente atualizado!\n");
            break;
        case OPCAO2:
            printf("\nDigite a nova enfermidade: ");
            scanf(" %[^\n]s", p->pacientes->enfermidade);
            LimpaBuffer();
            printf("Paciente atualizado!\n");
            break;
        case OPCAO3:
            printf("\nDigite a nova receita: ");
            scanf(" %[^\n]s", p->pacientes->receita);
            LimpaBuffer();
            printf("Paciente atualizado!\n");
            break;
        case OPCAO4:
            printf("Internado? (1-sim) (0-nao)\n");
            p->pacientes->internado = LeOpcao('0', '1');
            printf("Paciente atualizado!\n");
            break;
        case OPCAO5:
            printf("Digite '1' para editar o CPF ou '2' para editar o RG\n");
            opcao_documento = LeOpcao(OPCAO1, '2');
            if (opcao_documento == OPCAO1)
            {
                char cpf[15];
                lerCPF(cpf);
                strcpy(p->pacientes->documento.cpf, cpf);
            }
            else if (opcao_documento == OPCAO2)
            {
                char rg[12];
                lerRG(rg);
                strcpy(p->pacientes->documento.rg, rg);
            }
            break;
        default:
            printf("erro!\n");
            exit(1);
            break;
        }
    }
    else
    {
        printf("erro!\n");
        exit(1);
    }
    printf("\n");
}

void lista_libera(Listapacientes *l)
{
    Listapacientes *p = l;
    while (p != NULL)
    {
        Listapacientes *t = p->next;
        free(p);
        p = t;
    }
}

void lista_imprime(Listapacientes *l)
{
    Listapacientes *p = l->next;
    while (p != NULL)
    {
        printf("Nome: %s\n", p->pacientes->nome);
        printf("Enfermidade: %s\n", p->pacientes->enfermidade);
        printf("Receita: %s\n", p->pacientes->receita);
        printf("Internado: %s\n", p->pacientes->internado ? "Sim" : "Nao");
        if (strlen(p->pacientes->documento.cpf) == 14)
        {
            printf("CPF: %s\n", p->pacientes->documento.cpf);
        }
        else
        {
            printf("RG: %s\n", p->pacientes->documento.rg);
        }
        printf("\n");
        p = p->next;
    }
}

File: Hospital.py/src/paciente.py
import re

class Document:
    def __init__(self, cpf="", rg=""):
        self.cpf = cpf
        self.rg = rg

class Paciente:
    def __init__(self, nome="", documento=None, enfermidade="", internado=0, receita=""):
        self.nome = nome
        self.documento = documento if documento is not None else Document()
        self.enfermidade = enfermidade
        self.internado = internado
        self.receita = receita

class ListaPacientes:
    def __init__(self, paciente=Paciente()):
        self.paciente = paciente
        self.next = None
        self.prev = None


def corrige_nome(nome):
    # Verificar se o nome contém apenas letras e espaços
    nome_corrigido = ""
    palavras = nome.split()  # converter a string para uma lista de palavras

    # corrigir cada palavra e juntar de volta em uma única string
    for i in range(len(palavras)):
        palavra = palavras[i]
        palavra_corrigida = ""
        for j in range(len(palavra)):
            if j == 0:
                palavra_corrigida += palavra[j].upper()
            elif palavra[j-1] == " ":
                palavra_corrigida += palavra[j].upper()
            else:
                palavra_corrigida += palavra[j].lower()
        nome_corrigido += palavra_corrigida + " "

    return nome_corrigido.strip()  # remover espaços em branco no início e fim da string

def paciente_preenche():
    paciente = Paciente()
    nome = input("Digite o nome do paciente: ")
    paciente.nome = corrige_nome(nome)
    paciente.enfermidade = input("Digite a enfermidade do paciente: ")
    paciente.receita = input("Digite a receita para o paciente: ")
    while True:
        try:
            paciente.internado = int(input("Internado? (1-sim) (0-nao): "))
            if paciente.internado not in [0, 1]:
                raise ValueError
            break
        except ValueError:
            print("Opção inválida. Digite 1 para sim e 0 para não.")
    while True:
        try:
            opcao_documento = input("Digite '1' para cadastrar o CPF ou '2' para cadastrar o RG: ")
            if opcao_documento not in ['1', '2']:
                raise ValueError
            break
        except ValueError:
            print("Opção inválida. Digite 1 para CPF e 2 para RG.")
    if opcao_documento == '1':
        while True:
            cpf = input("Digite o CPF do paciente (XXX.YYY.ZZZ-SS): ")
            if re.match(r'^\d{3}\.\d{3}\.\d{3}-\d{2}$', cpf):
                paciente.documento.cpf = cpf
                break
            else:
                print("CPF inválido. Digite no formato XXX.YYY.ZZZ-SS.")

    elif opcao_documento == '2':
        while True:
            rg = input("Digite o RG do paciente (XXX.YYY.ZZZ): ")
            if re.match(r'^\d{3}\.\d{3}\.\d{3}$', rg):
                paciente.documento.rg = rg
                break
            else:
                print("RG inválido. Digite no formato XXX.YYY.ZZZ.")
        return paciente
        
def lista_add(self, paciente):
    # Encontrar a posição correta para inserir o novo paciente
    p = self.lista
    while p.next is not None and p.next.paciente.nome < paciente.nome:
        p = p.next

    # Criar um novo nó para o novo paciente
    novo_no = ListaPacientes(Paciente())
    novo_no.paciente = paciente

    # Inserir o novo paciente na posição correta
    novo_no.next = p.next
    novo_no.prev = p
    if p.next is not None:
        p.next.prev = novo_no
    p.next = novo_no
    
def lista_imprime(lista):
    p = lista.next
    while p != None:
        print("Nome:", p.paciente.nome)
        print("Enfermidade:", p.paciente.enfermidade)
        print("Receita:", p.paciente.receita)
        print("Internado:", "Sim" if p.paciente.internado else "Nao")
        if p.paciente.documento.cpf or p.paciente.documento.rg:
            if p.paciente.documento.cpf:
                print("CPF:", p.paciente.documento.cpf)
            else:
                print("RG:", p.paciente.documento.rg)
        print()
        p = p.next

def arquivo_add(lista, caminho):
    arquivo = open(caminho, "w")
    if arquivo == None:
        print("Erro ao abrir o arquivo.\n")
        exit(1)

    p = lista
    while p.next != None:
        paciente = p.next.paciente
        arquivo.write(f"Nome: {paciente.nome}\n")
        arquivo.write(f"Enfermidade: {paciente.enfermidade}\n")
        arquivo.write(f"Receita: {paciente.receita}\n")
        arquivo.write(f"Internado: {'Sim' if paciente.internado else 'Nao'}\n")
        if len(paciente.documento.cpf) == 14:
            arquivo.write(f"CPF: {paciente.documento.cpf}\n")
        else:
            arquivo.write(f"RG: {paciente.documento.rg}\n")
        arquivo.write("\n")
        p = p.next

    arquivo.close()
    return arquivo

def paciente_busca(lista, nome):
    p = lista
    while p != None:
        if p.paciente.nome == nome:
            return p
        p = p.next
    print(f"Paciente {nome} não encontrado na lista.")
    return None

def paciente_edita(p):
    print(f"\nNome: {p.nome}")
    print(f"Enfermidade: {p.enfermidade}")
    print(f"Receita: {p.receita}")
    print(f"Internado: {'Sim' if p.internado else 'Nao'}")
    if p.documento.cpf or p.documento.rg:
        if p.documento.cpf:
            print("CPF:", p.documento.cpf)
        else:
            print("RG:", p.documento.rg)
    while True:
                try:
                    cont = int(input("Deseja mesmo editar o paciente acima? (1-sim)(2-nao)\n"))
                    if cont not in [1, 2]:
                        raise ValueError
                    break
                except ValueError:
                    print("Opção inválida. Digite 1 para sim e 2 para não.")
    print()
    if cont == 1:
        opcao = int(input("\nO que voce deseja editar?\n1 - Nome\n2 - Enfermidade\n3 - Receita\n4 - Internado\n5 - Documento\nOpcao: "))
        if opcao == 1:
            nome = input("\nDigite o novo nome: ")
            p.nome = corrige_nome(nome)
        elif opcao == 2:
            p.enfermidade = input("\nDigite a nova enfermidade: ")
        elif opcao == 3:
            p.receita = input("\nDigite a nova receita: ")
        elif opcao == 4:
            p.internado = int(input("\nO paciente esta internado? (1-sim / 0-nao): "))
            while True:
                try:
                    p.internado = int(input("Internado? (1-sim) (0-nao): "))
                    if p.internado not in [0, 1]:
                        raise ValueError
                    break
                except ValueError:
                    print("Opção inválida. Digite 1 para sim e 0 para não.")
        elif opcao == 5:
            if p.documento.cpf or p.documento.rg:
                if p.documento.cpf:
                    while True:
                        cpf = input("Digite o novo CPF do paciente (XXX.YYY.ZZZ-SS): ")
                        if re.match(r'^\d{3}\.\d{3}\.\d{3}-\d{2}$', cpf):
                            p.documento.cpf = cpf
                            break
                        else:
                            print("CPF inválido. Digite no formato XXX.YYY.ZZZ-SS.")
                else:
                    while True:
                        rg = input("Digite o RG do pacientee (XXX.YYY.ZZZ): ")
                        if re.match(r'^\d{3}\.\d{3}\.\d{3}$', rg):
                            p.documento.rg = rg
                            break
                        else:
                            print("RG inválido. Digite no formato XXX.YYY.ZZZ.")     
        else:
            print("Opcao invalida!")
            return
    print("Paciente editado com sucesso!")